[TELCODOCS-257] Performing end-to-end tests for platform verification (#36024)
openshift/openshift-docs: modules/cnf-performing-end-to-end-tests-for-platform-verification.adoc

This allows your `kubeconfig` file to be consumed from inside the running container.

[WARNING]
====
You must run the latency tests in Discovery mode. The latency tests can change the configuration of your cluster if you do not run in Discovery mode.
====

> In {product-title} {product-version}, you can also run latency tests from the CNF-test container. The latency test allows you to validate the node tuning is sufficient for your workload.

Reviewer: s/he latency test allows you to validate the node/the latency test allows you to validate that the node

Three tools measure the latency of the system:

* `hwlatdetect`
* `cyclictest`
* `oslat`

Each tool has a specific use. Use the tools in sequence to achieve reliable test results.

. The `hwlatdetect` tool measures the baseline that the bare hardware can achieve. Before proceeding with the next latency test, ensure that the number measured by `hwlatdetect` meets the required threshold because hardware latency spikes cannot be fixed by operating system tuning.

.  The `cyclictest` tool verifies the timer latency after `hwlatdetect` passes validation. The `cyclictest` tool schedules a repeated timer and measures the difference between the desired and the actual trigger times. The difference can uncover basic issues with the tuning caused by interrupts or process priorities.

. The `oslat` tool behaves similarly to a CPU-intensive DPDK application and measures all the interruptions and disruptions to the busy loop that simulates CPU heavy data processing.

By default, the latency tests are disabled. To enable the latency test, you must add the `LATENCY_TEST_RUN` environment variable to the test invocation and set its value to `true`. For example, `LATENCY_TEST_RUN=true`.

> The test introduces the following environment variables:

Reviewer: I would verify that we use this structure, The best I can find is that we have examples starting with "Specify...".

I would revise this list to be "`The LATENCY_TEST_CPUS` flag (or whatever) specifies the number of CPUs"

* `LATENCY_TEST_RUNTIME` - Specifies the amount of time in seconds that the latency test must run.
* `CYCLICTEST_MAXIMUM_LATENCY` - Specifies the maximum latency in microseconds that all threads expect before waking up during the `cyclictest` run.
* `HWLATDETECT_MAXIMUM_LATENCY` - Specifies the maximum acceptable hardware latency in microseconds for the workload and operating system.
* `OSLAT_MAXIMUM_LATENCY` - Specifies the maximum acceptable latency in microseconds for the `oslat` test results.
* `MAXIMUM_LATENCY` - A unified variable you can apply for all tests.

[NOTE]
====
A variable that is specific to certain tests has precedence over the unified variable.
====

You can use the `ginkgo.focus` flag to run a specific test.

[id="cnf-performing-end-to-end-tests-running-hwlatdetect"]
=== Running hwlatdetect

To perform the `hwlatdetect`, run the following command:

[source, terminal]
----
$ docker run -v $(pwd)/:/kubeconfig -e KUBECONFIG=/kubeconfig/kubeconfig -e LATENCY_TEST_RUN=true -e DICOVERY_MODE=true -e LATENCY_TEST_RUNTIME=600 -e MAXIMUM_LATENCY=20  registry.redhat.io/openshift4/cnf-tests-rhel8:v4.9 /usr/bin/test-run.sh -ginko.focus=”hwladetect”
----

> The above command runs the `hwlatdetect` tool for 10 minutes (600 seconds). The test runs successfully when the maximum observed latency is lower than `MAXIMUM_LATENCY` (20 μs) and the command line displays `SUCCESS!`.

Reviewer: Comma after "(20 μs)"

k8s.io/klog.(*loggingT).output(0x5bed00, 0xc000000003, 0xc00010a0e0, 0x53ea81, 0x7, 0x33, 0x0)
	/remote-source/app/vendor/k8s.io/klog/klog.go:826 +0x35f
k8s.io/klog.(*loggingT).printf(0x5bed00, 0x3, 0x5082da, 0x33, 0xc000113f58, 0x2, 0x2)
	/remote-source/app/vendor/k8s.io/klog/klog.go:707 +0x153
k8s.io/klog.Fatalf(...)
	/remote-source/app/vendor/k8s.io/klog/klog.go:1276
main.main()
	/remote-source/app/cnf-tests/pod-utils/hwlatdetect-runner/main.go:51 +0x897
----
<1> The docker arguments provided by the user.
<2> The latency threshold configured by the user using the `MAX_LATENCY` or the `HWLATDETECT_MAX_LATENCY` environment variables.
<3> The maximum latency value measured during the test.

[id="cnf-performing-end-to-end-tests-running-cyclictest"]
=== Running cyclictest

To perform the `cyclictest`, run the following command:

[source, terminal]
----
$ docker run -v $(pwd)/:/kubeconfig -e KUBECONFIG=/kubeconfig/kubeconfig -e LATENCY_TEST_RUN=true -e DICOVERY_MODE=true -e LATENCY_TEST_CPUS=10 -e LATENCY_TEST_RUNTIME=600 -e MAXIMUM_LATENCY=20  registry.redhat.io/openshift4/cnf-tests-rhel8:v4.9 /usr/bin/test-run.sh -ginkgo.focus="cyclictest"
----

> The above command runs the `cyclictest` tool for 10 minutes (600 seconds). The test runs successfully when the maximum observed latency is lower than `MAXIMUM_LATENCY` (20 μs) and the command line displays `SUCCESS!`.

Reviewer: Comma after "(20 μs)"

This allows your `kubeconfig` file to be consumed from inside the running container.

[id="cnf-performing-end-to-end-tests-running-the-latency_tests_{context}"]
=== Running the latency tests

[WARNING]
====
You must run the latency tests in Discovery mode. The latency tests can change the configuration of your cluster if you do not run in Discovery mode.
====

> In {product-title} {product-version}, you can also run latency tests from the CNF-test container. The latency test allows you to set a latency limit so that you can determine performance, throughput, and latency.

Reviewer 2: ```suggestion
In {product-title} {product-version}, you can also run latency tests from the CNF-test container. The latency test allows you to validate the node tuning is good enough for your workloads.
```

Three tools measure the latency of the system:

* `hwlatdetect`
* `cyclictest`
* `oslat`

> Each tool has a specific use. Use the tools in sequence as the validation progresses.

Reviewer 2: You are right, the focus is explained lower in the doc though.

> . The `hwlatdetect` tool measures the baseline that the bare hardware can achieve. Before proceeding with the next latency test, ensure that the number measured by `hwlatdetect` meets the required threshold because latency spikes can yield unreliable results.

Reviewer 2: ```suggestion
. The `hwlatdetect` tool measures the baseline that the bare hardware can achieve. Before proceeding with the next latency test, ensure that the number measured by `hwlatdetect` meets the required threshold because hardware latency spikes cannot be fixed by operating system tuning.
```

> .  The `cyclictest` tool verifies the timer latency after `hwlatdetect` passes validation. The `cyclictest` tool schedules a repeated timer and measures the difference between the desired and the actual trigger times. The difference can uncover basic issues with the tuning caused by interrupts or process priorities.

Reviewer 2: Actually, if hwlatdetect gives bad results, cyclictest will also. Having good hwlatdetect run is a must to continue with any tuning.

. The `oslat` tool behaves similarly to a CPU-intensive DPDK application and measures all the interruptions and disruptions to the processing.

> By default, the latency tests are disabled. To enable the latency test, you must add the `LATENCY_TEST_RUN` variable and set its value to `true`. For example, `LATENCY_TEST_RUN=true`.

Reviewer 2: ```suggestion
By default, the latency tests are disabled. To enable the latency test, you must add the `LATENCY_TEST_RUN` environment variable to the test invocation and set its value to `true`. For example, `LATENCY_TEST_RUN=true`.
```

The test introduces the following environment variables:

* `LATENCY_TEST_CPUS` - Specifies the number of CPUs that the pod running the latency tests uses.
* `LATENCY_TEST_RUNTIME` - Specifies the amount of time in seconds that the latency test must run.
* `CYCLICTEST_MAXIMUM_LATENCY` - Specifies the maximum latency in microseconds that all threads expect before waking up during the `cyclictest` run.

> * `HWLATDETECT_MAXIMUM_LATENCY` - Specifies the maximum latency in microseconds that the hardware-firmware expects.

Reviewer 2: ```suggestion
* `HWLATDETECT_MAXIMUM_LATENCY` - Specifies the maximum hardware latency in microseconds that the workload and operating system can tolerate.
```

> * `OSLAT_MAXIMUM_LATENCY` - Specifies the maximum latency in microseconds that all buckets expect during the `oslat` test run.

Reviewer 2: ```suggestion
* `OSLAT_MAXIMUM_LATENCY` - Specifies the maximum latency in microseconds that the `oslat` test will consider acceptable.
```

    001 (us):	 19390720316 19141129810 20265099129 20280959461 19391991159 19119877333
    002 (us):	 5304 5249 5777 5947 6829 4971
    003 (us):	 28 14 434 47 208 21
    004 (us):	 1388 853 123568 152817 5576 0
    005 (us):	 207850 223544 103827 91812 227236 231563
    006 (us):	 60770 122038 277581 323120 122633 122357
    007 (us):	 280023 223992 63016 25896 214194 218395
    008 (us):	 40604 25152 24368 4264 24440 25115
    009 (us):	 6858 3065 5815 810 3286 2116
    010 (us):	 1947 936 1452 151 474 361
  ...
     Minimum:	 1 1 1 1 1 1 (us)
     Average:	 1.000 1.000 1.000 1.000 1.000 1.000 (us)
     Maximum:	 37 38 49 28 28 19 (us) <3>
     Max-Min:	 36 37 48 27 27 18 (us)
    Duration:	 599.667 599.667 599.667 599.667 599.667 599.667 (sec)
----
<1> The list of CPUs running the `oslat` command. Seven CPUs are provided through the `LATENCY_TEST_CPUS` variable. Only six CPUs are displayed in total because one is used to run the `oslat` tool.
<2> The user is notified about the measured latency and the configured latency.
<3> The maximum latency values in microseconds that are measured on each CPU.

[id="cnf-performing-end-to-end-tests-image-parameters_{context}"]

> == Image parameters

Reviewer 2: While reading the draft, this subtitle looked very weird. It was as if the latency tests could use different images. The "chapter" separation here is non-obvious.

Author: Unfortunately, this is a tool limitation. We can't control how the different heading levels render.

Reviewer 2: What about making Running latency tests a higher level section? It deserves it as it is a very specific testing case and it is big enough too. You can then move Latency testing right  before the Disconnected mode I think.

Author: I have to look into if moving Running latency tests to a higher level would break any links. 

`You can then move Latency testing right before the Disconnected mode I think.`
Do you mean Discovery mode section? Also, I'm not familiar with end-to-end testing, would that be a more logical place for running the latency tests?

Reviewer 2: Nope, I really meant the Disconnected section: https://deploy-preview-36024--osdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes?utm_source=github&utm_campaign=bot_dp#cnf-performing-end-to-end-tests-disconnected-mode_cnf-master

The issue right now is that Latency testing does not even show in the TOC.

Reviewer 2: Ah, it does.. nevermind. However all the sections around talk about end to end tests and the way to execute them.. and sadly most of that content does not apply to latency testing except discovery mode, disconnected mode and single node.

So I wonder if the current progression of sections is the best one

    Prerequisites
    Running the latency tests
    Image parameters
    Dry run
    Disconnected mode
    Discovery mode
    Running in a single node cluster
    Troubleshooting
    Test reports
    Impacts on the cluster

Author: I updated the headings after your comment, that's why the latency test shows up. I can make another quick update and move that section before Disconnected mode

Reviewer 2: I would probably propose something like this

    Prerequisites
    Dry run
    Disconnected mode
    Running in a single node cluster
    (all above applies to test environment)

    Impacts on the cluster -- the title is misleading, it also describes what tests are doing what
          Image parameters - This one is kind of weird, it should probably belong to the Impacts on the cluster  as it describes configuration for the described tests
    Discovery mode
    Running the latency tests
    (all are about tests and about the impact to the cluster)

    Troubleshooting
    Test reports
    (what to do with the results and when something goes wrong)

Reviewer 2: @fedepaol I think I would like to hear your opinion here, the section titles are a bit of a mess

Author: Let me get back to you about this!

Author: @MarSik Thank you for your patience, Martin! I updated the document with the proposed section order.  However, I'd also like to get it reviewed by the Perf/Scale team, just to make sure that I'm not stepping on any toes with the restructuring.

In the meantime, you can check it in the [preview](https://deploy-preview-36024--osdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes?utm_source=github&utm_campaign=bot_dp#cnf-performing-end-to-end-tests-running-the-latency_tests). Let me know what you think!

Reviewer 3: I'd rename image parameters to "Override images used by tests" if you want, but I don't feel it should be under the other section. It's totally unrelated and explains how to override the images used for pods.

Author: @MarSik @fedepaol 
If we put Image parameters as a subsection, we'll have to adjust the next subsection under that. In that case, we'll have the same problem as previously with the different test sections, where the headings didn't stand out enough.

I'd propose to only rename the headings:
- Impacts on the cluster > Impacts of tests on the cluster
- Image parameters > Override test image parameters

Let me know if I can proceed with these changes.

> [id="cnf-performing-end-to-end-tests-running-the-latency_tests"]

Reviewer 4: module ids should have _{context} at the end of them, so that they can pick up a context for reuse. See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#module-file-metadata

=== Running the latency tests

[WARNING]
====
You must run the latency tests in Discovery mode. The latency tests can change the configuration of your cluster if you do not run in Discovery mode.
====

In {product-title} {product-version}, you can also run latency tests from the CNF-test container. The latency test allows you to set a latency limit so that you can determine performance, throughput, and latency.

> Three tools are used for measuring the latency of the system:

Reviewer 4: Avoid passive voice. Three tools measure the latency of the system. You can use three tools to measure the latency of the system. You can use three tools sequentially to measure system latency. Etc.

* `hwlatdetect`
* `cyclictest`
* `oslat`

> Each tool has a specific use. All three tools are used in sequence as the validation progresses.

Reviewer 4: Again, avoid passive voice. Perhaps try an imperative tone. Use the tools in sequence as the validation progresses.

> . The `hwlatdetect` tool measures the baseline that is achievable by the bare hardware. Before proceeding with the next latency test, ensure that the number measured by `hwlatdetect` meets the required threshold because latency spikes can yield unreliable results.

Reviewer 4: Sometimes passive voice occurs when the subject comes after the verb. The `hwlatdetect` tool measures the baseline that the bare metal hardware can achieve.

.  The `cyclictest` tool verifies the timer latency after `hwlatdetect` passes validation. The `cyclictest` tool schedules a repeated timer and measures the difference between the desired and the actual trigger times. The difference can uncover basic issues with the tuning caused by interrupts or process priorities.

> . The `oslat` tool behaves similarly to a CPU-intensive DPDK application and measures all the interruptions and disruptions to the processing.

Reviewer 4: Maybe it's just me, but "to the processing" sounds somehow incomplete. The processing of what?

Reviewer 5: @MarSik Could you please clarify if we need to answer what the processing is. Thank you.

Author: @MarSik Please, clarify this question.

Reviewer 2: ```suggestion
. The `oslat` tool behaves similarly to a CPU-intensive DPDK application and measures all the interruptions and disruptions to the busy loop that simulates cpu heavy data processing.
```

Would this work?

By default, the latency tests are disabled. To enable the latency test, you must add the `LATENCY_TEST_RUN` variable and set its value to `true`. For example, `LATENCY_TEST_RUN=true`.

The test introduces the following environment variables:

> * `LATENCY_TEST_CPUS` - Specifies the amount of CPUs that is used by the pod which runs the latency tests.

Reviewer 4: Specifies the number of CPUs that the pod running the latency tests uses. Also, is this CPUs or cores? CPU pinning is sometimes specific to a core.

Reviewer 5: @MarSik Could you please clarify if this is CPUs or cores, per the peer review comment. Thank you.

* `LATENCY_TEST_RUNTIME` - Specifies the amount of time in seconds that the latency test must run.

> * `CYCLICTEST_MAXIMUM_LATENCY` - Specifies the maximum latency in microseconds that is expected from all threads to wake-up during the cyclictest run.

Reviewer 4: Specifies the maximum latency in microseconds that all threads expect before waking up during the `cyclictest` run.

> * `HWLATDETECT_MAXIMUM_LATENCY` - Specifies the maximum latency in microseconds that is expected from the hardware-firmware.

Reviewer 4: pecifies the maximum latency in microseconds that the hardware-firmware expects.

> * `OSLAT_MAXIMUM_LATENCY` - Specifies the maximum latency in seconds that is expected from all buckets during the oslat test run.

Reviewer 4: Specifies the maximum latency in seconds that all buckets expect during the `oslat` test run.

> * `MAXIMUM_LATENCY` - A unified variable that can be applied for all tests.

Reviewer 4: A unified latency variable you can apply for all tests.

[NOTE]
====
A variable that is specific to certain tests has precedence over the unified variable.
====

> You can use `ginkgo.focus` flag to run a specific test. 

Reviewer 4: You can use **the** `ginkgo.focus` flag to run a specific test.

k8s.io/klog.(*loggingT).output(0x5bed00, 0xc000000003, 0xc00010a0e0, 0x53ea81, 0x7, 0x33, 0x0)
	/remote-source/app/vendor/k8s.io/klog/klog.go:826 +0x35f
k8s.io/klog.(*loggingT).printf(0x5bed00, 0x3, 0x5082da, 0x33, 0xc000113f58, 0x2, 0x2)
	/remote-source/app/vendor/k8s.io/klog/klog.go:707 +0x153
k8s.io/klog.Fatalf(...)
	/remote-source/app/vendor/k8s.io/klog/klog.go:1276
main.main()
	/remote-source/app/cnf-tests/pod-utils/hwlatdetect-runner/main.go:51 +0x897
----
<1> The docker arguments provided by the user. 
//check docker run arguments?
<2> The latency threshold configured by the user using the `MAX_LATENCY` or the `HWLATDETECT_MAX_LATENCY` environment variables.
<3> The maximum latency value measured during the test.

[id="cnf-performing-end-to-end-tests-running-cyclictest"]
==== Running `cyclictest`

To perform the `cyclictest`, run the following command:

[source, terminal]
----
$ docker run -v $(pwd)/:/kubeconfig -e KUBECONFIG=/kubeconfig/kubeconfig -e LATENCY_TEST_RUN=true -e DICOVERY_MODE=true -e LATENCY_TEST_CPUS=10 -e LATENCY_TEST_RUNTIME=600 -e MAXIMUM_LATENCY=20  registry.redhat.io/openshift4/cnf-tests-rhel8:v4.9 /usr/bin/test-run.sh -ginkgo.focus="cyclictest"
----

> The above command runs the `cyclictest` tool for 10 minutes (600 seconds). The test runs successfully when the maximum observed latency is lower than `MAXIMUM_LATENCY` (20 us) and the command line displays `SUCCESS!`.

Reviewer 4: (20 us)? Is that supposed to be milliseconds (20 ms)? Or is it microseconds (20 μs)?

Reviewer 2: microseconds

# Histogram Overflows: 00000 00220 00159 00128 00202 00017 00069 00059 00045 00120
# Histogram Overflow at cycle number:
# Thread 0:
# Thread 1: 01142 01439 05305 … # 00190 others
# Thread 2: 20895 21351 30624 … # 00129 others
# Thread 3: 01143 17921 18334 … # 00098 others
# Thread 4: 30499 30622 31566 ... # 00172 others
# Thread 5: 145221 170910 171888 ...
# Thread 6: 01684 26291 30623 ...# 00039 others
# Thread 7: 28983 92112 167011 … 00029 others
# Thread 8: 45766 56169 56171 ...# 00015 others
# Thread 9: 02974 08094 13214 ... # 00090 others
----
<1> The docker arguments provided by the user. 
<2> The user is notified about the measured latency and the configured latency.
<3> The arguments for the `cyclictest` command.
<4> The maximum latencies measured on each thread.

[id="cnf-performing-end-to-end-tests-running-oslat"]
==== Running `oslat`

[source, terminal]
----
$ docker run -v $(pwd)/:/kubeconfig -e KUBECONFIG=/kubeconfig/kubeconfig -e LATENCY_TEST_RUN=true -e DICOVERY_MODE=true -e LATENCY_TEST_CPUS=7 -e LATENCY_TEST_RUNTIME=600 -e MAXIMUM_LATENCY=20  registry.redhat.io/openshift4/cnf-tests-rhel8:v4.9 /usr/bin/test-run.sh -ginkgo.focus="oslat"
----

> The above command runs the `cyclictest` tool for 10 minutes (600 seconds). The test runs successfully when the maximum observed latency is lower than `MAXIMUM_LATENCY` (20 us) and the command line displays `SUCCESS!`.

Reviewer 4: Same question. Is that micro μ seconds or milliseconds?

        Core:	 4 6 52 54 56 58
    CPU Freq:	 2096 2096 2096 2096 2096 2096 (Mhz)
    001 (us):	 19390720316 19141129810 20265099129 20280959461 19391991159 19119877333
    002 (us):	 5304 5249 5777 5947 6829 4971
    003 (us):	 28 14 434 47 208 21
    004 (us):	 1388 853 123568 152817 5576 0
    005 (us):	 207850 223544 103827 91812 227236 231563
    006 (us):	 60770 122038 277581 323120 122633 122357
    007 (us):	 280023 223992 63016 25896 214194 218395
    008 (us):	 40604 25152 24368 4264 24440 25115
    009 (us):	 6858 3065 5815 810 3286 2116
    010 (us):	 1947 936 1452 151 474 361
  ...
     Minimum:	 1 1 1 1 1 1 (us)
     Average:	 1.000 1.000 1.000 1.000 1.000 1.000 (us)
     Maximum:	 37 38 49 28 28 19 (us) <3>
     Max-Min:	 36 37 48 27 27 18 (us)
    Duration:	 599.667 599.667 599.667 599.667 599.667 599.667 (sec)
----
<1> The list of CPUs running the `oslat` command. Seven CPUs are provided through the `LATENCY_TEST_CPUS` variable. Only six CPUs are displayed in total because one is used to run the `oslat` tool.
<2> The user is notified about the measured latency and the configured latency.
<3> The maximum latency values in microseconds that are measured on each CPU.

> 

Reviewer 4: This block isn't rendering properly for me. The footnotes are showing up as text in a quote block, so something is wrong with the delimiter here.

Author: I ran an asciibinder build and seems to work for me, it looks good in the preview as well. Can I ask how did you try to render it?
